doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,365 @@
+
+mediacreation_version = openassetio_mediacreation.version()
+
+manager.initialize(settings, [mediacreation_version])
+```
+
+### Manager side
+```python
+ def initialize(self, managerSettings, schemaVersions, hostSession):
+   # Store versions for later use
+   self.__schemaVersions = schemaVersions
+```
+
+If you've read the principles above, you can likely see that this
+violates a few of them. Notably, the host has no method of supporting
+multiple versions, and all the onus of migrating between versions is on
+the manager, despite the fact that it may not understand it at all.
+
+Therefore, an out-of-data implementation will likely be a fair bit more
+complicated, and look more like a **negotiation**.
+
+### Negotiated
+
+In this conception, one side of the exchange (manager in this case),
+advertises a set of versions it supports. The other side then queries
+this list, selects the specific versions it would prefer to communicate
+in, then informs the other side which versions should be the medium
+of exchange going forward. Keep in mind, this is merely a passable
+representation of the concept of a negotiated, out of data approach,
+and not a final design.
+
+### Host side
+
+```python
+# During initial setup
+
+# In this example, supportedSchemaVersions returns a dictionary of the
+# provided schemas, mapped to a list of versions the manager supports
+# for those schemas
+# {
+#   "openassetio.mediacreation", [1,2,3]
+#   "companycorp.custom_traits", [1]
+# }
+supported_versions = manager.supportedSchemaVersions()
+
+# Verify that the trait packages we require are supported at all
+# by the manager
+if not openassetio_mediacreation.packageId() in supported_versions:
+  incompatible_with_manager_error()
+  return;
+
+if not companycorp_traits.packageId() in supported_versions:
+  incompatible_with_manager_error()
+  return;
+
+mediacreation_version = None
+companycorp_version = None
+
+# Verify the versions of the trait packages we are capable of supporting
+# are supported by the manager
+# This host is capable of working with mediacreation version 3 and 4,
+# and companycorp custom traits version 1. Given the choice, it would
+# prefer to work with mediacreation version 4.
+if 4 in supported_versions[openassetio_mediacreation.packageId()]:
+  mediacreation_version = 4
+else if 3 in supported_versions[openassetio_mediacreation.packageId()]:
+  mediacreation_version = 3
+else:
+  incompatible_with_manager_error()
+  return
+
+if 1 in supported_versions[companycorp_traits.packageId()]:
+  companycorp_version = 1
+else:
+  incompatible_with_manager_error()
+  return
+
+# Tell the manager up front what versions you are going to use.
+# These must be versions reported from `supportedSchemaVersions`.
+negotiated_schema_versions =
+{
+  openassetio_mediacreation.packageId():mediacreation_version,
+  companycorp_traits.packageId():companycorp_version
+}
+manager.informSchemaVersions(negotiated_schema_versions)
```

Comment:
I see what you mean here, but I feel putting it in has made the example a bit less clear, bringing questions of python style imports vs how this would work in c++, as well as just making the example a fair bit longer. I'm gonna leave this out.